src/asymmetric_crypto/ecies.rs

```diff
@@ -0,0 +1,313 @@
+
+use crate::{
+    asymmetric_crypto::DhKeyPair,
+    kdf,
+    reexport::rand_core::CryptoRngCore,
+    symmetric_crypto::{
+        aes_256_gcm_pure::{decrypt_combined, encrypt_combined, Aes256GcmCrypto},
+        nonce::NonceTrait,
+        Dem,
+    },
+    CryptoCoreError, KeyTrait,
+};
+
+/// Encrypts a message using Elliptic Curve Integrated Encryption Scheme
+/// (ECIES). This implementation uses SHAKE128 (XOF) as a KDF and AES256-GCM as
+/// a symmetric cipher.
+///
+/// This function encrypts a given message using ECIES with the provided
+/// receiver's public key. The encrypted message is returned as a
+/// `Result<Vec<u8>, CryptoCoreError>`.
+///
+/// # Arguments
+///
+/// * `rng: &mut R` - A mutable reference to a cryptographically secure random
+///   number generator.
+/// * `receiver_public_key: &DH::PublicKey` - A reference to the receiver's
+///   public key.
+/// * `msg: &[u8]` - A byte slice representing the message to be encrypted.
+/// * `shared_encapsulation_data: Option<&[u8]>` - An optional byte slice of
+///   shared encapsulation data.
+/// * `shared_authentication_data: Option<&[u8]>` - An optional byte slice of
+///   shared authentication data.
+///
+/// # Returns
+///
+/// * `Result<Vec<u8>, CryptoCoreError>` - The encrypted message as a `Vec<u8>`,
+///   or a `CryptoCoreError` if an error occurs.
+///
+/// # Example
+///
+/// ```
+/// use cosmian_crypto_core::{
+///    asymmetric_crypto::{curve25519::X25519KeyPair, DhKeyPair,ecies::ecies_encrypt},
+///    reexport::rand_core::SeedableRng,
+///    CsRng,
+/// };
+///
+/// let mut rng = CsRng::from_entropy();
+/// let key_pair: X25519KeyPair = X25519KeyPair::new(&mut rng);
+/// let msg = b"Hello, World!";
+///
+/// let _encrypted_message = ecies_encrypt::<
+///         CsRng,
+///         X25519KeyPair,
+///         { X25519KeyPair::PUBLIC_KEY_LENGTH },
+///         { X25519KeyPair::PRIVATE_KEY_LENGTH },
+///     >(
+///     &mut rng,
+///     &key_pair.public_key(),
+///     msg,
+///     None,
+///     None
+/// ).unwrap();
+/// ```
+#[allow(dead_code)]
+pub fn ecies_encrypt<R, DH, const PUBLIC_KEY_LENGTH: usize, const PRIVATE_KEY_LENGTH: usize>(
+    rng: &mut R,
+    receiver_public_key: &DH::PublicKey,
+    msg: &[u8],
+    shared_encapsulation_data: Option<&[u8]>,
+    shared_authentication_data: Option<&[u8]>,
+) -> Result<Vec<u8>, CryptoCoreError>
+where
+    R: CryptoRngCore,
+    DH: DhKeyPair<PUBLIC_KEY_LENGTH, PRIVATE_KEY_LENGTH>,
+    DH::PublicKey: From<DH::PrivateKey>,
+    for<'a, 'b> &'a DH::PublicKey: Add<&'b DH::PublicKey, Output = DH::PublicKey>
+        + Mul<&'b DH::PrivateKey, Output = DH::PublicKey>,
+    for<'a, 'b> &'a DH::PrivateKey: Add<&'b DH::PrivateKey, Output = DH::PrivateKey>
+        + Sub<&'b DH::PrivateKey, Output = DH::PrivateKey>
+        + Mul<&'b DH::PrivateKey, Output = DH::PrivateKey>
+        + Div<&'b DH::PrivateKey, Output = DH::PrivateKey>,
+{
+    // Generate an ephemeral key pair (r, R) where R = r.G
```

Comment:
You should quote the spec you are implementing in the documentation.
I don't think it is useful to paste each line of the spec in the comments if it is that simple to understand what you are doing. Maybe you can use the same naming as the spec? This may not be a good idea though as mathematiciens tend to pick obsure names.